### ISSUE TITLE
IPv6 Enabled (Dual Stack) Azure windows Kubernetes cluster pod creation failed due to improper handling of windows hostname in IPAM

### ISSUE THREAD

**Issue by jawaharjazz** [Azure/aks-engine#4623]
**We are facing "network plugin cni failed to set up pod "pod name" network: Failed to allocate v6 pool: Failed to delegate: Failed to allocate pool: Invalid address space" error while creating the Ipv6 enabled (Dual Stack) windows Kubernetes clusters in the Azure portal. This issue prevents from enabling IPv6 enabled windows clusters**

below log taken from ipamv6 log.txt file

2021/08/11 14:55:50 [1400] [ipam] Starting source ipv6NodeIpam.
2021/08/11 14:55:50 [1400] [ipam] Refreshing address source.
2021/08/11 14:55:50 [1400] [ipam] Failed to retrieve node using hostname: 1515k8s00000ABC with err nodes "1515k8s00000ABC" not found
2021/08/11 14:55:50 [1400] [ipam] Source refresh failed, err:nodes "1515k8s00000ABC" not found.
2021/08/11 14:55:50 [1400] [azure-vnet-ipamv6] Failed to allocate pool: Invalid address space.
2021/08/11 14:55:50 [1400] [cni-ipam] ADD command completed with result:<nil> err:Failed to allocate pool: Invalid address space.

**Steps To Reproduce**
       1. Create ACS Kubernetes cluster in Azure with dual stack enabled.
       2. Add new Windows VM into VMSS with upper case alphabets in windows hostname like 1515k8s00000ABC 
       3. Now check the pod status of the new node and it will fail with error like "Failed to allocate pool: Invalid address space".

**Expected behavior**
      1. Pod should spin up in new node without any issues.

**AKS Engine version**
https://github.com/Azure/aks-engine/releases/tag/v0.62.1

**Kubernetes version**
  Kubernetes:1.20.5

**Additional context**
     I have checked the Kubernetes and azure cni soure code and found the possible root cause for this issue. kubectl get node returns node name with lower case even though host name has upper case alphabets, and we can verify the same with below image and **describe node 1515k8s00000ABC** is failed with error like "Error from server (not found): 1515k8s00000ABC not found"

![image](https://user-images.githubusercontent.com/67739968/130104232-4a31edf3-616e-4f23-a3e8-a8b8a365bf46.png)

below code is responsible for refreshing the address source in IPAM

https://github.com/Azure/azure-container-networking/blob/b63e52328c80b448121dac519b303e76ac16120b/ipam/manager.go#L238 

https://github.com/Azure/azure-container-networking/blob/b63e52328c80b448121dac519b303e76ac16120b/ipam/ipv6Ipam.go#L168 
        `kubeNode, err := source.kubeClient.CoreV1().Nodes().Get(context.TODO(), source.nodeHostname, metav1.GetOptions{})`

here source.nodeHostname passed as a parameter to get the node information and from the error log we can find the value of the source.nodeHostname which is 1515k8s00000ABC. This node get call gets failed and logged the previous error "Failed to retrieve node using hostname: 1515k8s00000ABC with err nodes "1515k8s00000ABC" not found" (check ipamv6log.txt error log)

below code is responsible for kubectl describe 

https://github.com/kubernetes/kubectl/blob/0f88fc6b598b7e883a391a477215afb080ec7733/pkg/describe/describe.go#L3469 
                        ` node, err := d.CoreV1().Nodes().Get(context.TODO(), name, metav1.GetOptions{})`
          here the name we have passed in the kubectl describe command passed as an argument. We are getting a successful response for describe node when hostname passed in lowercase otherwise not. Both kubectl describe and ipam hits the same end point to get the node information. I am suspecting the root cause could be the source.nodeHostname because it has upper case suffix in host name and due to this only node get call failed in IPAM.

**Comment by matmerr** [comment]
@tamilmani1989

**Comment by tamilmani1989** [comment]
@jawaharjazz as i mentioned in icm, please open the issue with aks-engine : https://github.com/Azure/aks-engine/issues 

closing this as its not azure cni issue

**Comment by marosset** [comment]
@aramase FYI

**Comment by marosset** [comment]
IPv6 / dualstack support on windows requires at least K8s v1.19 and [aks-engine v0.61.0](https://github.com/Azure/aks-engine/releases/tag/v0.61.0).

Based on info in the issue both the K8s and aks-engine version you are using do not have the necessary support.
Please try using newer versions of both.

**Comment by tamilmani1989** [comment]
@marosset the issue here is if nodename has uppercase alphabet at end, then `kubectl describe nodes` is failing with node not found error. This looks like case sensitiveness is not handled in windows k8s. Is this issue fixed in latest versions that you specified?

**Comment by marosset** [comment]
> @marosset the issue here is if nodename has uppercase alphabet at end, then `kubectl describe nodes` is failing with node not found error. This looks like case sensitiveness is not handled in windows k8s. Is this issue fixed in latest versions that you specified?

I'm not sure.
@aramase did you run into any issues with Windows host names being uppercase during your testing?

**Comment by jawaharjazz** [comment]
@marosset sorry for the incorrect information. I have previously tried on latest versions only but mentioned wrong version. I can reproduce the issue Kubernetes 1.20.5 version and aksEngineVersion v0.62.1 version.

below log is the azure-vnet-ipam6.log.
2021/08/24 05:56:37 [6232] [ipam] Starting source ipv6NodeIpam.
2021/08/24 05:56:37 [6232] [ipam] Refreshing address source.
2021/08/24 05:56:37 [6232] [ipam] Failed to retrieve node using hostname: 8126k8s0000000G with err nodes "8126k8s0000000G" not found
2021/08/24 05:56:37 [6232] [ipam] Source refresh failed, err:nodes "8126k8s0000000G" not found.
2021/08/24 05:56:37 [6232] [azure-vnet-ipamv6] Failed to allocate pool: Invalid address space.
2021/08/24 05:56:37 [6232] [cni-ipam] ADD command completed with result:<nil> err:Failed to allocate pool: Invalid address space.
2021/08/24 05:56:37 [6232] [cni-ipam] Plugin stopped

**Comment by aramase** [comment]
> @aramase did you run into any issues with Windows host names being uppercase during your testing?

I haven't run in to this issue so far. @jawaharjazz Few questions for you:

1. Are you able to reproduce this consistently?
2. Which all Kubernetes version have you tried this with?
3. The issue is only when the node name has uppercase?

**Comment by jawaharjazz** [comment]
1. Yes
2. 1.20.5
3. Yes. Issue can be reproducible when kubectl get node returns node names in lowercase but the hostname command returns hostname in upper case.

Below image contains hostname response of the windows node and you can see that the node name ends with upper case char.

![node hostname](https://user-images.githubusercontent.com/67739968/130763665-b4cb6cf1-1632-4ab8-946f-d7c9c8d458e2.png)

![get nodes ](https://user-images.githubusercontent.com/67739968/130763895-5e4a69ce-22a5-4235-a105-31efb910f5e2.png)

**Comment by jsturtevant** [comment]
I tracked this down to the way IPAM is set up in IPV6:

https://github.com/Azure/azure-container-networking/blob/0cf6ed8c3de2ef9afb3b024616ce0c932992b75d/ipam/ipv6Ipam.go#L62

VMSS sets the HOSTNAME to the name of the VMSS instance, which has capital letters in it.  Kubelet registers the node name with HOSTNAME but lowercases hence the miss match it: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/util/node/node.go#L69

This is a restriction in kubernetes due that can't be changed in a backwards compatible way: RFC 952/1123 allows uppercase characters, but IsDNS1123Subdomain does not · Issue #71140 · kubernetes/kubernetes (github.com)

The fix would be to lowercase the value in IPAM controller in azurecni.  I looked but didn't see a similar type of thing for ipv4 which is why I don't think we see this with IPV4 only.  @Tamilmani should be able to confirm.
